In [4]:
#P4: Presenting Data Stories with Jupyter
#Visualizing Bitcoin Data
#Inline Display allows Juypter to the display (plot) within the notebook.
%matplotlib inline



#import statistics
from datetime import date
import numpy as np

import pandas as pd
import matplotlib.pyplot as plt



```markdown
# Bitcoin Closing Prices for the Last 20 Days

Here is a list of Bitcoin closing prices over the last 20 days:

1. **Sep 20, 2024**: $60,768.89  
2. **Sep 19, 2024**: $61,046.56  
3. **Sep 18, 2024**: $59,028.27  
4. **Sep 17, 2024**: $58,388.7  
5. **Sep 16, 2024**: $61,677.3  
6. **Sep 15, 2024**: $60,264.2  
7. **Sep 14, 2024**: $59,570.2  
8. **Sep 13, 2024**: $54,786.9  
9. **Sep 12, 2024**: $57,025.6  
10. **Sep 11, 2024**: $53,998.2  
11. **Sep 10, 2024**: $49,486.9  
12. **Sep 09, 2024**: $57,346.9  
13. **Sep 08, 2024**: $61,242.3  
14. **Sep 07, 2024**: $64,538.3  
15. **Sep 06, 2024**: $65,328.7  
16. **Sep 05, 2024**: $66,544.5  
17. **Sep 04, 2024**: $67,067.8  
18. **Sep 03, 2024**: $66,776.8  
19. **Sep 02, 2024**: $63,253.5  
20. **Sep 01, 2024**: $67,586.4  

This period includes fluctuations between $49,486.9 and $67,586.4.
```

```markdown
Bitcoin volatility is a measure of how much the price of Bitcoin fluctuates over a period of time. A common way to calculate volatility is by using the **standard deviation** or **annualized volatility** of Bitcoin's daily returns. Here's how you can calculate it:

### Formula for Volatility

1. **Calculate daily returns**:
   \[
   R_t = \frac{P_t}{P_{t-1}} - 1
   \]
   where:
   - \( R_t \) is the return at time \( t \),
   - \( P_t \) is the closing price of Bitcoin at time \( t \),
   - \( P_{t-1} \) is the closing price at time \( t-1 \).

2. **Calculate the average return** \( \mu \):
   \[
   \mu = \frac{1}{N} \sum_{t=1}^{N} R_t
   \]
   where \( N \) is the total number of days.

3. **Calculate the variance** \( \sigma^2 \):
   \[
   \sigma^2 = \frac{1}{N} \sum_{t=1}^{N} (R_t - \mu)^2
   \]

4. **Calculate the standard deviation (volatility)** \( \sigma \):
   \[
   \sigma = \sqrt{\sigma^2}
   \]

5. **Annualize the volatility** (optional):
   \[
   \sigma_{\text{annual}} = \sigma \times \sqrt{252}
   \]
   Here, 252 represents the average number of trading days in a year.

### Steps for Calculation
1. Get the historical Bitcoin closing prices.
2. Compute the daily returns using the formula above.
3. Find the average and variance of the returns.
4. The square root of the variance gives you the volatility.
5. Multiply by \( \sqrt{252} \) if you want the annualized volatility.

source:  OpenAI's ChatGPT. (2024, 09 22). Bitcoin Volatility Calculation Explanation. Retrieved from ChatGPT: https://chatgpt.com/c/66f02c9d-44c8-8004-97a3-995398033b07


```

In [5]:
#Closing Bitcoin Prices from Sept 1st to Sept 20th
bitcoin_closing_prices = [
    {"date": "Sep 20, 2024", "price": 60768.89},
    {"date": "Sep 19, 2024", "price": 61046.56},
    {"date": "Sep 18, 2024", "price": 59028.27},
    {"date": "Sep 17, 2024", "price": 58388.7},
    {"date": "Sep 16, 2024", "price": 61677.3},
    {"date": "Sep 15, 2024", "price": 60264.2},
    {"date": "Sep 14, 2024", "price": 59570.2},
    {"date": "Sep 13, 2024", "price": 54786.9},
    {"date": "Sep 12, 2024", "price": 57025.6},
    {"date": "Sep 11, 2024", "price": 53998.2},
    {"date": "Sep 10, 2024", "price": 49486.9},
    {"date": "Sep 09, 2024", "price": 57346.9},
    {"date": "Sep 08, 2024", "price": 61242.3},
    {"date": "Sep 07, 2024", "price": 64538.3},
    {"date": "Sep 06, 2024", "price": 65328.7},
    {"date": "Sep 05, 2024", "price": 66544.5},
    {"date": "Sep 04, 2024", "price": 67067.8},
    {"date": "Sep 03, 2024", "price": 66776.8},
    {"date": "Sep 02, 2024", "price": 63253.5},
    {"date": "Sep 01, 2024", "price": 67586.4}
]


In [6]:



# Calculate daily returns
returns = [bitcoin_closing_prices[i]['price'] / bitcoin_closing_prices[i-1]['price'] - 1 for i in range(1, len(bitcoin_closing_prices))]
# Calculate volatility (standard deviation of returns)
volatility = np.std(returns)

# Annualized volatility
annualized_volatility = volatility * np.sqrt(252)

print("Volatility:", volatility)
print("Annualized Volatility:", annualized_volatility)


Volatility: 0.05748071783562524
Annualized Volatility: 0.9124781074472358


'''markdown
Context:
The output you received contains two important figures related to the volatility of Bitcoin's price:

1. **Volatility: 0.0575 (5.75%)**  
   This value represents the **daily volatility** of Bitcoin, calculated based on price fluctuations over the observed period. A value of 5.75% means that, on average, the price of Bitcoin changes by around 5.75% from one day to the next. It reflects how much the price fluctuates on a daily basis.

2. **Annualized Volatility: 0.9125 (91.25%)**  
   This value represents the **annualized volatility** of Bitcoin, assuming these daily fluctuations are consistent throughout the year. A value of 91.25% means that if the current pattern continues, the price of Bitcoin could vary by about 91.25% over the course of a year. This indicates a high level of risk or uncertainty in Bitcoin's price movements.

In summary, Bitcoin experiences significant day-to-day price changes (5.75%), and if those changes continue, its price could fluctuate by about 91.25% over the course of a year, indicating high volatility.

source:  OpenAI's ChatGPT. (2024, 09 22). explain the output data in plane English Volatility: 0.05748071783562524Annualized Volatility: 0.9124781074472358. Retrieved from ChatGPT: https://chatgpt.com/c/66f02c9d-44c8-8004-97a3-995398033
'''

'''markdown
Here are some interesting ways to graph the Bitcoin data you calculated:

### 1. **Daily Returns over Time (Line Chart)**  
   - **Graph**: A line chart showing the daily returns.
   - **Why**: This graph will highlight how Bitcoin's price fluctuates each day. It can help identify patterns of high volatility or stability over time.
   - **X-axis**: Dates
   - **Y-axis**: Daily returns (percentage change)

### 2. **Price vs. Time (Line Chart)**  
   - **Graph**: A line chart of Bitcoin’s daily closing prices over the 20-day period.
   - **Why**: It shows the overall trend in Bitcoin’s price, helping to visually spot increases or decreases in value. This could give insights into whether Bitcoin is trending upward or downward.
   - **X-axis**: Dates
   - **Y-axis**: Closing prices

### 3. **Distribution of Daily Returns (Histogram)**  
   - **Graph**: A histogram of the daily returns.
   - **Why**: This graph shows the distribution of Bitcoin’s daily price changes. It helps visualize the frequency of different return ranges (e.g., how often the return was between -1% and 1%).
   - **X-axis**: Return intervals (e.g., -5% to +5%)
   - **Y-axis**: Frequency of returns in each interval

### 4. **Volatility over Time (Rolling Standard Deviation) (Line Chart)**  
   - **Graph**: A line chart showing the rolling volatility over time (e.g., a 5-day rolling standard deviation).
   - **Why**: This can help visualize how the volatility changes over time. For instance, are there periods where the price is more stable vs. more volatile?
   - **X-axis**: Dates
   - **Y-axis**: Rolling volatility (e.g., 5-day or 10-day)

### 5. **Annualized Volatility (Bar Chart)**  
   - **Graph**: A bar chart showing the calculated annualized volatility.
   - **Why**: This will show the magnitude of Bitcoin’s annual volatility relative to the daily volatility. It’s useful for comparing the risk over time, especially when analyzing multiple time periods.
   - **X-axis**: Time period (e.g., "Annualized Volatility")
   - **Y-axis**: Value of volatility

### 6. **Price Change and Daily Returns Comparison (Dual-Axis Line Chart)**  
   - **Graph**: A dual-axis chart showing both price change (on one axis) and daily returns (on the other axis).
   - **Why**: This can give a holistic view of both the absolute price changes and relative daily returns, helping you see how percentage returns relate to price movements.
   - **X-axis**: Dates
   - **Left Y-axis**: Closing prices
   - **Right Y-axis**: Daily returns (percentage)

These graphs would provide clear, actionable insights into Bitcoin’s volatility and price trends over time.

source:  OpenAI's ChatGPT. (2024, 09 22). what would be interesting to graph in the data points calculated.  what kind of a graph. Retrieved from ChatGPT: https://chatgpt.com/c/66f02c9d-44c8-8004-97a3-995398033
'''

In [ ]:

# Convert the list of dictionaries into a DataFrame
bitcoin_closing_prices_df = pd.DataFrame(bitcoin_closing_prices)

# Convert the 'date' column to datetime format and set it as the index
bitcoin_closing_prices_df['date'] = pd.to_datetime(bitcoin_closing_prices_df['date'])
bitcoin_closing_prices_df.set_index('date', inplace=True)



# Convert returns to a DataFrame for easier plotting
returns_df = pd.DataFrame(returns, index=bitcoin_closing_prices_df.index[1:], columns=['Daily Return'])

# Plotting the daily returns
plt.figure(figsize=(14, 7))
plt.plot(returns_df.index, returns_df['Daily Return'], label='Daily Returns', color='blue')
plt.title('Bitcoin Daily Returns Over Time')
plt.xlabel('Date')
plt.ylabel('Daily Returns (%)')
plt.axhline(0, color='red', linewidth=0.8, linestyle='--')  # Line at 0% for reference
plt.legend()
plt.grid()
plt.show()
